routing/payment_lifecycle.go

```diff
@@ -119,413 +200,287 @@ func (p *paymentLifecycle) resumePayment() ([32]byte, *route.Route, error) {
+		// order,
+		//   1. check timeout.
+		//   2. request route.
+		//   3. create HTLC attempt.
+		//   4. send HTLC attempt.
+		//   5. collect HTLC attempt result.
+		//
+		// Before we attempt any new shard, we'll check to see if
+		// either we've gone past the payment attempt timeout, or the
+		// router is exiting. In either case, we'll stop this payment
+		// attempt short. If a timeout is not applicable, timeoutChan
+		// will be nil.
+		if err := p.checkTimeout(); err != nil {
+			return exitWithErr(err)
 		}
 
-		// If we either reached a terminal error condition (but had
-		// active shards still) or there is no remaining value to send,
-		// we'll wait for a shard outcome.
-		wait, err := payment.NeedWaitAttempts()
+		// Now decide the next step of the current lifecycle.
+		step, err := p.decideNextStep(payment)
 		if err != nil {
 			return exitWithErr(err)
 		}
 
-		if wait {
-			// We still have outstanding shards, so wait for a new
-			// outcome to be available before re-evaluating our
-			// state.
-			if err := shardHandler.waitForShard(); err != nil {
-				return exitWithErr(err)
-			}
-			continue lifecycle
-		}
-
-		// Before we attempt any new shard, we'll check to see if
-		// either we've gone past the payment attempt timeout, or the
-		// router is exiting. In either case, we'll stop this payment
-		// attempt short. If a timeout is not applicable, timeoutChan
-		// will be nil.
-		select {
-		case <-p.timeoutChan:
-			log.Warnf("payment attempt not completed before " +
-				"timeout")
-
-			// By marking the payment failed with the control
-			// tower, no further shards will be launched and we'll
-			// return with an error the moment all active shards
-			// have finished.
-			saveErr := p.router.cfg.Control.FailPayment(
-				p.identifier, channeldb.FailureReasonTimeout,
-			)
-			if saveErr != nil {
-				return exitWithErr(saveErr)
-			}
+		switch step {
+		// Exit the for loop and return below.
+		case stepExit:
+			break lifecycle
 
+		// Continue the for loop and skip the rest.
+		case stepSkip:
 			continue lifecycle
 
-		case <-p.router.quit:
-			return exitWithErr(ErrRouterShuttingDown)
+		// Continue the for loop and proceed the rest.
+		case stepProceed:
 
-		// Fall through if we haven't hit our time limit.
+		// Unknown step received, exit with an error.
 		default:
+			err = fmt.Errorf("unknown step: %v", step)
+			return exitWithErr(err)
 		}
 
-		// Create a new payment attempt from the given payment session.
-		rt, err := p.paySession.RequestRoute(
-			ps.RemainingAmt, remainingFees,
-			uint32(ps.NumAttemptsInFlight),
-			uint32(p.currentHeight),
-		)
+		// Now request a route to be used to create our HTLC attempt.
+		rt, err := p.requestRoute(ps)
 		if err != nil {
-			log.Warnf("Failed to find route for payment %v: %v",
-				p.identifier, err)
-
-			routeErr, ok := err.(noRouteError)
-			if !ok {
-				return exitWithErr(err)
-			}
+			return exitWithErr(err)
```

Comment:
One thing which came to my mind today is, why are we not marking this payment failed or at least with a status that we can later remove/delete it? There are some other places in this loop as well, but I am wondering why we just exit the loop if a critical err happens without bringing the payment into a garbage collectable state?

Reply:
This is similar to why we handle routing error and critical error differently - if we know for sure it's a routing error, we can be sure this payment won't succeed. For critical error, it means non-routing related systems have failed - it could be as serious as a db corruption or simply the node is shutting down by a user. If the node is then restarted, this payment will be resumed. Depending on its timeout value, it may be failed instantly on startup, or continue the lifecycle to collect HTLC attempt results.

So in short, if a critical error is returned, it means we need manual intervention to fix the case. Once it's fixed, this payment can still be tried.

Reply:
Thanks for the explanation. So a follow-up question I have then: So currently it does look like when restarting an inflight payment, that we do not check that we actually are passed timeout of the invoice or ?:
```
_, _, err := r.sendPayment(
				0, payment.Info.PaymentIdentifier, 0,
				paySession, shardTracker,
			)
```
Wondering if we should do this, or is this caught somewhere else down the process? Could have the invoice stored, so it wouldn't be a problem or ? Because when a payment fails, I think most noderunner (rebalances) will not notice anyway and restart their node only rarely so the possibility of the payment already outdated is realistic imo ?

Moreover I think one could add the manual trigger to restart the payment, if a critical err returned, just making it more comfortable for the user ? 

All the above point are non-blocking tho.

Reply:
Good question! I believe we should continue track and fix this in #5461